src/overlays/actors/ovl_Shot_Sun/z_shot_sun.c

```diff
@@ -32,27 +30,154 @@ const ActorInit Shot_Sun_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Shot_Sun/func_80973960.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Shot_Sun/ShotSun_Update.s")
+void ShotSun_Init(Actor* thisx, PlayState* play) {
+    s32 pad;
+    ShotSun* this = THIS;
+
+    if ((SHOTSUN_GET_FF(thisx) == SHOTSUN_40) || (SHOTSUN_GET_FF(thisx) == SHOTSUN_41)) {
+        this->unused = 0;
+        this->actor.flags |= ACTOR_FLAG_10;
+        this->actor.flags |= ACTOR_FLAG_2000000;
+        this->actionFunc = func_809738D0;
+        this->actor.flags |= ACTOR_FLAG_8000000;
+    } else {
+        Collider_InitCylinder(play, &this->collider);
+        Collider_SetCylinder(play, &this->collider, &this->actor, &sCylinderInit);
+        this->actionFunc = ShotSun_UpdateHyliaSun;
+        this->actor.flags &= ~ACTOR_FLAG_1;
+    }
+}
+
+void ShotSun_Destroy(Actor* thisx, PlayState* play) {
+    ShotSun* this = THIS;
+
+    if ((SHOTSUN_GET_FF(thisx) != SHOTSUN_40) && (SHOTSUN_GET_FF(thisx) != SHOTSUN_41)) {
+        Collider_DestroyCylinder(play, &this->collider);
+    }
+}
+
+void ShotSun_SpawnFairy(ShotSun* this, PlayState* play2) {
+    s32 params = SHOTSUN_GET_FF(&this->actor);
+    s32 fairyType = 0;
+    PlayState* play = play2;
+
+    if (this->timer > 0) {
+        this->timer--;
+    } else {
+        ActorCutscene_Stop(this->actor.cutscene);
+        switch (params) {
+            case SHOTSUN_40:
+                fairyType = 7;
+                break;
+            case SHOTSUN_41:
+                fairyType = 7;
+                break;
+        }
+        Actor_Spawn(&play->actorCtx, play, ACTOR_EN_ELF, this->actor.home.pos.x, this->actor.home.pos.y,
+                    this->actor.home.pos.z, 0, 0, 0, fairyType);
+        Actor_MarkForDeath(&this->actor);
+    }
+}
+
+void ShotSun_TriggerFairy(ShotSun* this, PlayState* play) {
+    if ((this->actor.cutscene == -1) || ActorCutscene_GetCanPlayNext(this->actor.cutscene)) {
+        if (this->actor.cutscene != -1) {
+            ActorCutscene_Start(this->actor.cutscene, &this->actor);
+        }
+        this->actionFunc = ShotSun_SpawnFairy;
+        this->timer = 50;
+        Actor_Spawn(&play->actorCtx, play, ACTOR_DEMO_KANKYO, this->actor.home.pos.x, this->actor.home.pos.y,
+                    this->actor.home.pos.z, 0, 0, 0, 0x11);
+        Audio_PlaySfxAtPos(&this->actor.projectedPos, NA_SE_EV_TRE_BOX_APPEAR);
+    } else {
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+    }
+}
+
+void func_809738D0(ShotSun* this, PlayState* play) {
+    s32 params = SHOTSUN_GET_FF(&this->actor);
+
+    if (play->msgCtx.ocarinaMode == 3) {
+        switch (play->msgCtx.lastPlayedSong) {
+            case OCARINA_SONG_STORMS:
+                if (params == SHOTSUN_41) {
+                    this->actionFunc = ShotSun_TriggerFairy;
+                    play->msgCtx.ocarinaMode = 4;
+                }
+                break;
+            case OCARINA_SONG_SUNS:
+                if (params == SHOTSUN_40) {
```

Comment:
Indeed, you can name the params based on the song required, `SHOTSUN_FAIRY_SPAWNER_STORMS` and `SHOTSUN_FAIRY_SPAWNER_SUNS`.